File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.maxD = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        def getDepth(root):
            if not root:
                return 0
            
            left = 1 + getDepth(root.left)
            right = 1 + getDepth(root.right)

            self.maxD = max(self.maxD, left-1 + right-1)

            return max(right, left)

        getDepth(root)

        return self.maxD
        

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int goodNodes = 0;

    public void dfs(TreeNode root, int pathMax) {
        if (root == null) {
            return;
        }

        pathMax = Math.max(pathMax, root.val);
        if (pathMax == root.val) {
            goodNodes++;
        }

        dfs(root.left, pathMax);
        dfs(root.right, pathMax);
    }
    public int goodNodes(TreeNode root) {
        int pathMax = root.val;

        dfs(root, pathMax);
        return goodNodes;
        
    }
}
